hw/drivers/adc/adc_nrf51/src/adc_nrf51.c

```diff
@@ -0,0 +1,409 @@
+    cc_cfg = &nrf_adc_chan->config.config;
+    reference = cc_cfg->reference | (cc_cfg->external_reference << ADC_CONFIG_EXTREFSEL_Pos);
+
+    /* Set the resolution and reference voltage for this channel to
+    * enable conversion functions.
+    */
+    switch (cc_cfg->resolution) {
+        case NRF_ADC_CONFIG_RES_8BIT:
+            res = 8;
+            break;
+        case NRF_ADC_CONFIG_RES_9BIT:
+            res = 9;
+            break;
+        case NRF_ADC_CONFIG_RES_10BIT:
+            res = 10;
+            break;
+        default:
+            assert(0);
+    }
+
+    switch (reference) {
+        case NRF_ADC_CONFIG_REF_VBG:
+            refmv = 1200; /* 1.2V for NRF51 */
+            break;
+        case NRF_ADC_CONFIG_REF_EXT_REF0:
+            refmv = init_adc_config->nadc_refmv0;
+            break;
+        case NRF_ADC_CONFIG_REF_EXT_REF1:
+            refmv = init_adc_config->nadc_refmv1;
+            break;
+        case NRF_ADC_CONFIG_REF_SUPPLY_ONE_HALF:
+            refmv = init_adc_config->nadc_refmv_vdd / 2;
+            break;
+        case NRF_ADC_CONFIG_REF_SUPPLY_ONE_THIRD:
+            refmv = init_adc_config->nadc_refmv_vdd / 3;
+            break;
+        default:
+            assert(0);
+    }
+
+    /* Adjust reference voltage for gain. */
+    switch (cc_cfg->input) {
+        case NRF_ADC_CONFIG_SCALING_INPUT_FULL_SCALE:
+            break;
+        case NRF_ADC_CONFIG_SCALING_INPUT_ONE_THIRD:
+            refmv *= 3;
+            break;
+        case NRF_ADC_CONFIG_SCALING_INPUT_TWO_THIRDS:
+            refmv = (refmv * 3) / 2;
+            break;
+        case NRF_ADC_CONFIG_SCALING_SUPPLY_ONE_THIRD:
+            refmv = refmv * 3;
+            break;
+        case NRF_ADC_CONFIG_SCALING_SUPPLY_TWO_THIRDS:
+            refmv = (refmv * 3) / 2;
+            break;
+        default:
+            break;
+    }
+
+    /* Store these values in channel definitions, for conversions to
+     * milivolts.
+     */
+    dev->ad_chans[cnum].c_res = res;
+    dev->ad_chans[cnum].c_refmv = refmv;
+    dev->ad_chans[cnum].c_configured = 1;
+    dev->ad_chans[cnum].c_cnum = cnum;
+
+    return (0);
+err:
+    return (rc);
+}
+
+/**
+ * Set buffer to read data into.  Implementation of setbuffer handler.
+ * Sets both the primary and secondary buffers for DMA.
+ */
+static int
+nrf51_adc_set_buffer(struct adc_dev *dev, void *buf1, void *buf2,
+                     int buf_len)
+{
+    int rc;
+
+    if (dev == NULL || buf1 == NULL) {
+        rc = OS_EINVAL;
+        goto err;
+    }
+
+    /* Convert overall buffer length, into a total number of samples which
+     * Nordic APIs expect.
+     */
+    buf_len /= sizeof(nrf_adc_value_t);
+
+    rc = nrfx_adc_buffer_convert((nrf_adc_value_t *) buf1, buf_len);
+    if (rc != NRFX_SUCCESS) {
+        goto err;
+    }
+
+    return (0);
+err:
+    return (rc);
+}
+
+//WHY DOES THIS EXIST! why doesnt it work without it!
```

Comment:
Heh. I dont understand why calling nrfx_adc_buffer_convert would 'release' the buffer, but when I remove this it breaks....